main/src/ca/uwaterloo/flix/language/phase/unification/FastBoolUnification.scala

```diff
@@ -0,0 +1,1694 @@
+
+      case Term.And(_, vars, rest) => SortedSet.empty[Int] ++ vars.map(_.x) ++ rest.flatMap(_.freeVars)
+
+      case Term.Or(ts) => ts.foldLeft(SortedSet.empty[Int])(_ ++ _.freeVars)
+    }
+
+    /**
+      * Returns the size of `this` term.
+      *
+      * The size is the number of constants, variables, and connectives in the term.
+      *
+      * For example, `size(x /\ y) = 1` and `size(x /\ not y) = 2`.
+      */
+    final def size: Int = this match {
+      case Term.True => 0
+      case Term.False => 0
+      case Term.Cst(_) => 1
+      case Term.Var(_) => 0
+      case Term.Not(t) => t.size + 1
+      case Term.And(csts, vars, rest) =>
+        // We need a connective for each constant, variable, and term minus one.
+        // We then add the size of all the sub-terms in `rest`.
+        ((csts.size + vars.size + rest.size) - 1) + rest.map(_.size).sum
+      case Term.Or(ts) => ts.map(_.size).sum + (ts.length - 1)
+    }
+
+    /**
+      * Returns a human-readable representation of `this` term.
+      */
+    override def toString: String = this match {
+      case Term.True => formatter.red("true")
+      case Term.False => formatter.red("false")
+      case Term.Cst(c) => formatter.blue(s"c$c")
+      case Term.Var(x) => s"x$x"
+      case Term.Not(f) => f match {
+        case Term.Var(x) => s"¬x$x"
+        case _ => s"¬($f)"
+      }
+      case Term.And(csts, vars, rest) => s"(${(csts.toList ++ vars.toList ++ rest).mkString(" ∧ ")})"
+      case Term.Or(ts) => s"(${ts.mkString(" ∨ ")})"
+    }
+
+  }
+
+  object Term {
+
+    /**
+      * The TRUE symbol.
+      */
+    case object True extends Term
+
+    /**
+      * The FALSE symbol.
+      */
+    case object False extends Term
+
+    /**
+      * Represents an uninterpreted constant ("rigid variable").
+      *
+      * Note: We assume that constants and variables are disjoint.
+      */
+    case class Cst(c: Int) extends Term
+
+    /**
+      * Represents a Boolean variable ("flexible variable").
+      *
+      * Note: We assume that constants and variables are disjoint.
+      */
+    case class Var(x: Int) extends Term
+
+    /**
+      * Represents the negation of the term `t`.
+      */
+    case class Not(t: Term) extends Term
+
+    /**
+      * Represents a conjunction of terms.
+      *
+      * We use a clever representation where we have a conjunction of constants, variables, and then sub-terms.
+      *
+      * For example, the conjunction: `x7 /\ (not x2) /\ c1 /\ x4` is represented as: `Set(c1), Set(x4, x7), List((not x2))`.
+      *
+      * This representation is key to efficiency because the equations we solve are heavy on conjunctions.
+      */
+    case class And(csts: Set[Term.Cst], vars: Set[Term.Var], rest: List[Term]) extends Term {
+      // We ensure that `rest` cannot contain constants and variables.
+      // Once the code is better tested, we can remove these assertions.
+      assert(!rest.exists(_.isInstanceOf[Term.Cst]))
```

Comment:
maybe put them behind the `verify` bool?

Reply:
I want to have them enable by default. Later I will remove them. (Recall verify is very expensive and cannot be enabled by default).